working/macros/feature-specification.md

```diff
@@ -550,6 +550,116 @@ This does have two interesting and possibly unexpected consequences:
 
 **TODO**: Define the API for adding metadata to existing declarations (#1931).
 
+## Aspect Macros
+
+**NOTE**: This feature is a work in progress, and has many TODOs.
+
+An AspectMacro is used by another macro, to generate shared helper code which
+will not be duplicated across every library where that helper was needed.
+
+Concretely, let's consider the following example:
+
+```dart
+@fromJsonExtension
+class A {}
+
+@fromJsonExtension
+class B {
+  final A a;
+
+  B(this.a);
+}
+
+// Generated extension from the macro on A
+extension AFromJson on Map<String, Object?> {
+  A toA();
+}
+
+// Generated extension from the macro on B
+extension BFromJson on Map<String, Object?> {
+  A toA(); // This also needs to generate a helper for converting `A`!
+
+  B toB();
+}
+```
+
+Both of these macro applications need to generate the same shared code for
+converting a Map into an instance of `A` (since `B` has a field of that type),
+but they can't actually _see_ what each other has generated.
+
+In the worst case, this would result in them both outputting conflicting
+members, and in the best case they would end up generating duplicate code,
+causing unnecessary code bloat.
+
+Aspect Macros exist to solve this problem, by allowing macros to share these
+implementations. This is done by associating a single, canonical declaration
+with an Aspect Macro Instance + Declaration pair, and giving back an opaque
+identifier for the resulting declaration.
+
+### Applying an Aspect Macro
+
+Aspect Macros can only be applied by other macros (including aspect macros),
+while the macro is running. They do this through the following api:
+
+```dart
+Future<Identifier> applyAspect(AspectMacro macro, Declaration declaration);
+```
+
+This API does not promise to immediately invoke `macro` on `declaration`, it
+only returns an opaque `Identifier`, which refers to the declaration that will
+be produced by running `macro` on `declaration`. It returns a `Future` because
+all `Identifier` objects must come from the host compiler, so there is some
+async communication that has to happen.
+
+When a macro implementation receives a call to `applyAspect`, it must register
+that `macro` should be _eventually_ applied to `declaration`. Any two identical
+calls to `applyAspect` must always return an "equal" `Identifier` object,
```

Comment:
By "identical calls" you mean where the parameters are the same, I assume? Do the AspectMacro instances need to be identical, or just instances of the same class?

It might be worth considering allowing some other kind of data or context parameters in order to parameterize what the macro does.

Reply:
I see really two viable options here:

1.   Say that all `AspectMacro`s have to be _equal_.
    - We may want to enforce that they override `==` and `hashCode` if they have any fields?
2.   Don't allow `AspectMacro`s to have fields at all.

Fwiw, the later approach is basically the bazel approach, aspects are not configurable. If you want different behavior you have to create a different aspect.

Given that we also have to send these across the wire, they have to be serializable (and deserializable) as well. I am inclined to just go with approach 2 and say they can't have fields at all (but they can have getters, which allows for customized versions of the aspect).

In general we probably don't want to easily encourage customizable aspects because then they can't share code.